Code chanllenge for SuperFormular

MarceKalEl opened the pull request:
Hi all! 

Description:
This PR introduces a series of enhancements and structural refinements to our Flutter project, acknowledging the limitations posed by external dependencies and optimizing our architecture and feature set. A key focus has been placed on adhering to Clean Architecture principles, albeit with a tailored approach to suit the already established repository patterns within the project. Below are the key highlights and rationale behind these changes:

Constant JSON Usage: In response to Yelp API's rate limiting, we've opted to utilize a constant JSON for data handling. This approach ensures that our development and testing processes remain uninterrupted by external API constraints, allowing us to simulate a stable data environment.

Clean Architecture Adaptation: While our project's scope and existing repository layout did not necessitate a full-blown Clean Architecture implementation, we've made strides in aligning with its principles where feasible. This adaptation ensures a scalable and maintainable codebase, facilitating easier future enhancements and feature additions.

BLoC and Feature-First Paradigm: The presentation layer has been restructured to leverage the BLoC pattern, emphasizing a feature-first approach. This reorganization enhances readability, maintainability, and the ease of feature scalability, aligning with Flutter's reactive programming model.

Splash Screen Introduction: To improve the user experience and provide a seamless app startup, a splash screen has been implemented. This addition not only enhances the aesthetic appeal but also prepares the app's state and resources in the background, leading to a smoother transition for the user.

Functional Programming with Oxidized: Embracing functional programming concepts, we've integrated Oxidized into our project. This inclusion allows for more expressive and concise code, particularly in handling success and error states, thereby improving overall code quality and readability.

Future Improvements: Acknowledging the project's potential for further enhancement, we've identified areas such as internal database management, string management, and API endpoint expansion as opportunities for growth. The current limitations posed by Yelp's API have underscored the need for a more robust solution to external data handling and integration.

This PR lays the groundwork for a more robust, scalable, and maintainable application, setting a solid foundation for future developments and enhancements. Your feedback and contributions to these changes are highly valued as we continue to refine and expand our project's capabilities.
I am looking forward to your feedback and suggestions on how we can further improve this implementation. Your insights are invaluable to enhancing the project's quality and functionality.

Warm regards.

alberto-sf commented on lib/core/app_init.dart:
```diff
@@ -0,0 +1,9 @@
+import 'package:flutter_dotenv/flutter_dotenv.dart';
+import 'package:restaurantour/core/helpers/hive_helper.dart';
+
+class AppInit {
```
why did you create a class with a single `static` member instead of just having a top-level function? like

```dart
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:restaurantour/core/helpers/hive_helper.dart';

Future<void> initializeApp() async {
  await dotenv.load(fileName: ".env");
  await HiveHelper().init();
}
```

MarceKalEl replied:
I understand your point. However, despite of the fact that it is a small test app, I considered that if the app evolves, some other methods might be added like
initializeFirebase
setupAnalytics
andOtherSimilars
so, on the Appinit I can call the methods in order, like:
```
class Appinit (
static Future<void> initializeApp() async {
await dotenv.load (fileName: "env"); await HiveHelper).init);
await initializeFirebase: setupAnalytics);
｝

```
I know that there are other ways to do so, yet I found it more readeable. In addition, I consider that it may be usefull in some other cases or scenarios, like I dont't know. Maybe, user switching? instead restart whole app or all services, just some specific ones can be called.

alberto-sf commented on lib/core/helpers/dio_helper.dart:
```diff
@@ -0,0 +1,16 @@
+import 'package:flutter_dotenv/flutter_dotenv.dart';
+
+class DioHelper {
+  static final Dio _dio = Dio(
```
same question here 🙂  why not just using the equivalent

```dart
import 'package:dio/dio.dart';
import 'package:flutter_dotenv/flutter_dotenv.dart';

final dio = Dio(
  BaseOptions(...)
):
```

MarceKalEl replied:
I encapsulated Dio in a static class for flexibility and centralization of network functionalities. In my opinion,  this setup allows adding helper methods, custom interceptors, and adapting other features impacting dependent code.
If another way can be used, i guess I just a matther of old habits i have. :P

alberto-sf commented on lib/core/models/Failure.dart:
```diff
@@ -0,0 +1,27 @@
+import 'package:equatable/equatable.dart';
+
+abstract class Failure extends Equatable {
```
this could have been a [`sealed`](https://dart.dev/language/class-modifiers#sealed) class

alberto-sf commented on lib/features/home_page/children/all_restaurant/presenter/bloc/all_restaurant/all_restaurant_bloc.dart:
```diff
@@ -0,0 +1,53 @@
+    emit(
+      const LoadingState(),
+    );
+    final yelpRepo = yelpRepository;
```
why not using `yelpRepository` directly?

MarceKalEl replied:
totally right!  dunno why I missed that :P  sorry

MarceKalEl replied:
fixed now.

alberto-sf commented on lib/features/home_page/children/all_restaurant/presenter/page/all_restaurants_tab.dart:
```diff
@@ -0,0 +1,85 @@
+          );
+        }
+      },
+      child: const _Body(),
```
very nice 😄 I like when `const` constructors and reusable widgets are created!

MarceKalEl replied:
thanks I appreciate it :D 

![image](https://github.com/Superformula/flutter_test/assets/22898121/f351c6a2-78cf-4bb3-9ccd-c55b68cb48e0)

alberto-sf commented on lib/features/home_page/children/favorite_restaurants/domain/entities/category_entity.dart:
```diff
@@ -0,0 +1,9 @@
+class CategoryEntity {
```
this is ok but with Dart 3 records this could just have been

```dart
typedef CategoryEntity = ({String title, String alias})
```

which also has `==`, hash code and to String defined for you (something that your class has not)

MarceKalEl replied:
Thank you for the suggestion. You're absolutely right. This indeed simplifies the implementation of CategoryEntity, I can tell that now is cleaner and less verbose syntax compared to traditional class-based implementations.

alberto-sf reviewed (commented):
Thank you for your time! 😄 

I really like how you split widgets and use `const` constructors. I also like how you've written tests and created the widget tree.

I have some comments about some Dart coding choices I'd like you to clarify!

MarceKalEl commented:
Hi Alberto! :D Thank you very much for the feedback. I have already
replied to some of your questions, and addressed some comments you
left.

Hope you are having a nice weekend.
Cheers!

El vie, 1 mar 2024 a las 16:29, Alberto Miola
***@***.***>) escribió:
>
> @alberto-sf commented on this pull request.
>
> Thank you for your time! 😄
>
> I really like how you split widgets and use const constructors. I also like how you've written tests and created the widget tree.
>
> I have some comments about some Dart coding choices I'd like you to clarify!
>
> ________________________________
>
> In lib/core/app_init.dart:
>
> > @@ -0,0 +1,9 @@
> +import 'package:flutter_dotenv/flutter_dotenv.dart';
> +import 'package:restaurantour/core/helpers/hive_helper.dart';
> +
> +class AppInit {
>
> why did you create a class with a single static member instead of just having a top-level function? like
>
> import 'package:flutter_dotenv/flutter_dotenv.dart';
> import 'package:restaurantour/core/helpers/hive_helper.dart';
>
> Future<void> initializeApp() async {
>   await dotenv.load(fileName: ".env");
>   await HiveHelper().init();
> }
>
> ________________________________
>
> In lib/core/helpers/dio_helper.dart:
>
> > @@ -0,0 +1,16 @@
> +import 'package:dio/dio.dart';
> +import 'package:flutter_dotenv/flutter_dotenv.dart';
> +
> +class DioHelper {
> +  static final Dio _dio = Dio(
>
> same question here 🙂 why not just using the equivalent
>
> import 'package:dio/dio.dart';
> import 'package:flutter_dotenv/flutter_dotenv.dart';
>
> final dio = Dio(
>   BaseOptions(...)
> ):
>
> ________________________________
>
> In lib/core/models/Failure.dart:
>
> > @@ -0,0 +1,27 @@
> +import 'package:equatable/equatable.dart';
> +
> +abstract class Failure extends Equatable {
>
> this could have been a sealed class
>
> ________________________________
>
> In lib/features/home_page/children/all_restaurant/presenter/bloc/all_restaurant/all_restaurant_bloc.dart:
>
> > +  AllRestaurantBloc({required this.hiveHelper, required this.yelpRepository})
> +      : super(AllRestaurantInitial()) {
> +    on<IniEvent>(_onInitEvent);
> +  }
> +
> +  final HiveHelper hiveHelper;
> +  final YelpRepository yelpRepository;
> +
> +  Future<void> _onInitEvent(
> +    IniEvent event,
> +    Emitter<AllRestaurantState> emit,
> +  ) async {
> +    emit(
> +      const LoadingState(),
> +    );
> +    final yelpRepo = yelpRepository;
>
> why not using yelpRepository directly?
>
> ________________________________
>
> In lib/features/home_page/children/all_restaurant/presenter/page/all_restaurants_tab.dart:
>
> > +  const _Page();
> +
> +  @override
> +  Widget build(BuildContext context) {
> +    return BlocListener<AllRestaurantBloc, AllRestaurantState>(
> +      listener: (context, state) {
> +        if (state is ErrorState) {
> +          ScaffoldMessenger.of(context).showSnackBar(
> +            const SnackBar(
> +              content: Text('Something went wrong, please come back later.'),
> +              duration: Duration(seconds: 3),
> +            ),
> +          );
> +        }
> +      },
> +      child: const _Body(),
>
> very nice 😄 I like when const constructors and reusable widgets are created!
>
> ________________________________
>
> In lib/features/home_page/children/favorite_restaurants/domain/entities/category_entity.dart:
>
> > @@ -0,0 +1,9 @@
> +class CategoryEntity {
>
> this is ok but with Dart 3 records this could just have been
>
> typedef CategoryEntity = ({String title, String alias})
>
> which also has ==, hash code and to String defined for you (something that your class has not)
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>



-- 
Marcelo Sánchez Melgar || Flutter Developer || MEFN Fullstack developer
Ph:+(591) 75293301

diegog-sf commented:
Hello @MarceKalEl , I'm having an issue running your project, 
```
Error (Xcode): No file or variants found for asset: .env.
```

could you update the readme with instructions on how to run the app?

MarceKalEl commented:
sure updating now.

El mié, 6 mar 2024 a las 16:20, Diego Garcia ***@***.***>)
escribió:

> Hello @MarceKalEl <https://github.com/MarceKalEl> , I'm having an issue
> running your project,
>
> Error (Xcode): No file or variants found for asset: .env.
>
> could you update the readme with instructions on how to run the app?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Superformula/flutter_test/pull/14#issuecomment-1981715555>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFOWLSIPAQYCDV4DSGER3EDYW53CXAVCNFSM6AAAAABDM7YMTCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOBRG4YTKNJVGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
*Marcelo Sánchez Melgar || Flutter Developer || MEFN Fullstack developer*


*Ph:+(591) 75293301*
